compiler/src/openxla/compiler/nvgpu/Conversion/ConvertHLOToCUDNN.cpp

```diff
@@ -201,34 +212,56 @@ static FailureOr<ConvLayouts> matchConv(stablehlo::ConvolutionOp op,
   return {{*inputLayout, *kernelLayout, *outputLayout}};
 }
 
-// Matches any clamp 'op' which can be converted to relu and outlines it into
-// a cudnn.graph op.
-// TODO(chsigg): extend this to match a set of ops (determined through a cost
-// model) to run as a single graph.
-static LogicalResult outlineClamp(stablehlo::ClampOp op,
-                                  PatternRewriter& rewriter) {
-  if (failed(matchRelu(op, rewriter))) return failure();
-  auto type = getCudnnTensorType(op.getType(), Layout::NCHW);
-  FunctionType funcType = rewriter.getFunctionType(type, type);
-  return outlineToGraph(op.getResult(), op.getOperand(), funcType, rewriter);
+// Matches defining ops starting from 'result' and returns the graph inputs.
+static LogicalResult outlineToGraph(TensorValue result,
+                                    PatternRewriter& rewriter) {
+  // The layout of the first convolution output, if we encounter such an op.
+  // Eventually, this should be replaced by layout propagation.
+  std::optional<Layout> layout;
+  SmallVector<TensorValue> arguments = {result};
+  for (size_t i = 0; i < arguments.size();) {
+    Operation* defOp = arguments[i].getDefiningOp();
+    if (defOp == nullptr) {
+      ++i;
+      continue;
+    }
+    if (auto op = dyn_cast<stablehlo::ClampOp>(defOp)) {
+      if (succeeded(matchRelu(op, rewriter))) {
+        arguments[i] = op.getOperand();
+      } else {
+        ++i;
+      }
+      continue;
+    }
+    if (auto op = dyn_cast<stablehlo::ConvolutionOp>(defOp)) {
+      auto layouts = matchConv(op, rewriter);
+      if (succeeded(layouts) && (!layout || layout == layouts->output)) {
+        layout = layouts->output;
+        arguments[i] = op.getLhs();
+        arguments.push_back(op.getRhs());
+      } else {
+        ++i;
+      }
+      continue;
+    }
+    ++i;
+  }
+  if (arguments.size() == 1 && arguments.front() == result) return failure();
+  auto getType = [&](TensorValue operand) -> Type {
+    return getCudnnTensorType(operand.getType(), layout.value_or(Layout::NCHW));
+  };
+  FunctionType funcType = rewriter.getFunctionType(
+      to_vector(map_range(arguments, getType)), getType(result));
+  return outlineToGraph(result, arguments, funcType, rewriter);
```

Comment:
At some point outlining has to check that all intermediates has one use (or better no uses outside of an outlined graph), I don't think cuDNN supports materializing intermediate results into buffers.

Reply:
Good point, I forgot about that. I will fix this in a follow-up.